cumulus/pallets/parachain-system/src/validate_block/implementation.rs

```diff
@@ -205,55 +205,21 @@ where
 		let mut overlay = OverlayedChanges::default();
 
 		parent_header = block.header().clone();
-		let inherent_data = extract_parachain_inherent_data(&block);
-
-		validate_validation_data(
-			&inherent_data.validation_data,
-			relay_parent_number,
-			relay_parent_storage_root,
-			&parachain_head,
-		);
-
-		// We don't need the recorder or the overlay in here.
-		run_with_externalities_and_recorder::<B, _, _>(
-			&backend,
-			&mut Default::default(),
-			&mut Default::default(),
-			|| {
-				let relay_chain_proof = crate::RelayChainStateProof::new(
-					PSC::SelfParaId::get(),
-					inherent_data.validation_data.relay_parent_storage_root,
-					inherent_data.relay_chain_state.clone(),
-				)
-				.expect("Invalid relay chain state proof");
-
-				#[allow(deprecated)]
-				let res = CI::check_inherents(&block, &relay_chain_proof);
-
-				if !res.ok() {
-					if log::log_enabled!(log::Level::Error) {
-						res.into_errors().for_each(|e| {
-							log::error!("Checking inherent with identifier `{:?}` failed", e.0)
-						});
-					}
-
-					panic!("Checking inherents failed");
-				}
-			},
-		);
 
-		run_with_externalities_and_recorder::<B, _, _>(
-			&execute_backend,
-			// Here is the only place where we want to use the recorder.
-			// We want to ensure that we not accidentally read something from the proof, that was
-			// not yet read and thus, alter the proof size. Otherwise we end up with mismatches in
-			// later blocks.
-			&mut execute_recorder,
-			&mut overlay,
-			|| {
-				E::execute_block(block);
-			},
-		);
+		run_with_validation_params(&mut validation_params, || {
```

Comment:
Personally I find it a bit hard to follow the flow because of the global variables.

Here the idea is to do this as early as possible for efficiency ?

Or for example could we do

```
				let vfp =
					crate::ValidationData::<PSC>::get().expect("Validation data not present");
				validation_params.validate_vfp(&vfp);
```

after executing the block ?